inference-engine/src/mkldnn_plugin/nodes/non_max_suppression.cpp

```diff
@@ -135,110 +213,202 @@ class NonMaxSuppressionImpl: public ExtLayerBase {
+        int *valid_outputs = nullptr;
+        if (outputs.size() > 2)
+            valid_outputs = outputs[NMS_VALIDOUTPUTS]->buffer().as<int *>() +
+                            outputs[NMS_VALIDOUTPUTS]->getTensorDesc().getBlockingDesc().getOffsetPadding();
+
+        const SizeVector &boxesStrides = inputs[NMS_BOXES]->getTensorDesc().getBlockingDesc().getStrides();
+        const SizeVector &scoresStrides = inputs[NMS_SCORES]->getTensorDesc().getBlockingDesc().getStrides();
 
         // boxes shape: {num_batches, num_boxes, 4}
         // scores shape: {num_batches, num_classes, num_boxes}
-        int num_batches = static_cast<int>(scores_dims[0]);
-        int num_classes = static_cast<int>(scores_dims[1]);
-        std::vector<filteredBoxes> fb;
-
-        for (int batch = 0; batch < num_batches; batch++) {
-            float *boxesPtr = boxes + batch * boxesStrides[0];
-            for (int class_idx = 0; class_idx < num_classes; class_idx++) {
-                float *scoresPtr = scores + batch * scoresStrides[0] + class_idx * scoresStrides[1];
-                std::vector<std::pair<float, int> > scores_vector;
-                for (int box_idx = 0; box_idx < num_boxes; box_idx++) {
-                    if (scoresPtr[box_idx] > score_threshold)
-                        scores_vector.push_back(std::make_pair(scoresPtr[box_idx], box_idx));
-                }
+        const int &num_batches = static_cast<int>(scores_dims[0]);
+        const int &num_classes = static_cast<int>(scores_dims[1]);
+
+        std::vector<filteredBoxes> filtBoxes(max_output_boxes_per_class * num_batches * num_classes);
+
+        auto coeff = [&](float iou) {
+            const float weight = std::exp(scale * iou * iou);
+            return iou <= iou_threshold ? weight : 0.0f;
+        };
+
+        auto less = [](const boxInfo& l, const boxInfo& r) {
+            return l.score < r.score || ((l.score == r.score) && (l.idx > r.idx));
+        };
+
+        parallel_for2d(num_batches, num_classes, [&](int B, int C) {
+            for (int batch = B; batch < num_batches; batch++) {
+                const float *boxesPtr = boxes + batch * boxesStrides[0];
+                for (int class_idx = C; class_idx < num_classes; class_idx++) {
+                    std::vector<filteredBoxes> fb;
+                    const float *scoresPtr = scores + batch * scoresStrides[0] + class_idx * scoresStrides[1];
+
+                    std::priority_queue<boxInfo, std::vector<boxInfo>, decltype(less)> sorted_boxes(less);
+                    for (int box_idx = 0; box_idx < num_boxes; box_idx++) {
+                        if (scoresPtr[box_idx] > score_threshold)
+                            sorted_boxes.emplace(boxInfo({scoresPtr[box_idx], box_idx, 0}));
+                    }
+
+                    if (sorted_boxes.size() > 0) {
+                        while (fb.size() < static_cast<size_t>(max_output_boxes_per_class) && !sorted_boxes.empty()) {
```

Comment:
Why not to declare max_output_boxes_per_class as size_t instead of casting? It is good approach to avoid cast operations inside loops. They may significantly affect 'fast' layers performance.

Reply:
fixed